### PR TITLE
Closes #85 Integrate Elasticsearch for Fast Search Access to Startups and Projects

### LINKED ISSUE #85: [Elasticsearch] Integrate Elasticsearch for Fast Search Access to Startups and Projects
#### Objective
To enhance the search performance and capabilities in the project, we aim to integrate Elasticsearch. This will enable quick, complex search queries on `Startup` and `Project` models, supporting features like keyword searches, filtering, and sorting.

---

### Sub-Tasks

1. **Setup and Installation**

   - **Install Elasticsearch**:
     - Run Elasticsearch locally or use a Docker container for development. For production, you may choose a managed Elasticsearch service.
   - **Install Required Libraries**:
     - Install `django-elasticsearch-dsl` and `django-elasticsearch-dsl-drf` libraries in your Django project.
     ```bash
     pip install django-elasticsearch-dsl django-elasticsearch-dsl-drf
     ```

2. **Define Elasticsearch Settings**

   - **Update Django Settings**:
     - Configure Elasticsearch settings in `settings.py`.
     ```python
     ELASTICSEARCH_DSL = {
         'default': {
             'hosts': 'localhost:9200'
         },
     }
     ```

3. **Create Elasticsearch Documents**

   - **Define Documents for Models**:
     - Create Elasticsearch document definitions for `Startup` and `Project` models. These documents will map model fields to Elasticsearch fields.
     - Create a `documents.py` file in each relevant app (e.g., `startups` and `projects`).
     ```python
     # startups/documents.py
     from django_elasticsearch_dsl import Document, Index
     from .models import Startup

     startup_index = Index('startups')

     @startup_index.doc_type
     class StartupDocument(Document):
         class Django:
             model = Startup
             fields = [
                 'company_name',
                 'description',
                 'funding_stage',
                 'location',
                 'industries',
             ]
     ```

   - **Repeat for Projects**:
     ```python
     # projects/documents.py
     from django_elasticsearch_dsl import Document, Index
     from .models import Project

     project_index = Index('projects')

     @project_index.doc_type
     class ProjectDocument(Document):
         class Django:
             model = Project
             fields = [
                 'title',
                 'description',
                 'status',
                 'required_amount',
                 'startup',
             ]
     ```

4. **Synchronize Data with Elasticsearch**

   - **Index Existing Data**:
     - Run a management command to index existing data in Elasticsearch.
     ```bash
     python manage.py search_index --rebuild
     ```

   - **Set Up Django Signals for Syncing**:
     - Use Django signals to keep Elasticsearch indexes updated when data in the database changes.
     - Add signals in `signals.py` in each app to listen for `post_save` and `post_delete` events on `Startup` and `Project`.

5. **Create DRF Views for Search Functionality**

   - **Build Search Views**:
     - Implement views in `views.py` that leverage Elasticsearch to perform searches on `Startup` and `Project`.
     - Use `django-elasticsearch-dsl-drf` to create DRF views for search endpoints.
     ```python
     # startups/views.py
     from django_elasticsearch_dsl_drf.viewsets import DocumentViewSet
     from .documents import StartupDocument
     from .serializers import StartupDocumentSerializer

     class StartupSearchViewSet(DocumentViewSet):
         document = StartupDocument
         serializer_class = StartupDocumentSerializer
         filter_backends = [
             FilteringFilterBackend,
             OrderingFilterBackend,
         ]
         search_fields = ('company_name', 'description')
         filter_fields = {
             'funding_stage': 'exact',
             'location': 'exact',
         }
     ```

   - **Repeat for Projects**.

6. **Create Serializers for Elasticsearch Documents**

   - **Define Elasticsearch Document Serializers**:
     - Create serializers for `StartupDocument` and `ProjectDocument`.
     ```python
     # startups/serializers.py
     from rest_framework import serializers
     from .documents import StartupDocument

     class StartupDocumentSerializer(serializers.ModelSerializer):
         class Meta:
             model = StartupDocument
             fields = '__all__'
     ```

7. **Update URLs**

   - **Add URLs for Search Endpoints**:
     - Create URL routes for `StartupSearchViewSet` and `ProjectSearchViewSet`.
     ```python
     # startups/urls.py
     from django.urls import path
     from .views import StartupSearchViewSet

     urlpatterns = [
         path('search/', StartupSearchViewSet.as_view({'get': 'list'}), name='startup-search'),
     ]
     ```

8. **Testing**

   - **Write Unit Tests**:
     - Test the search functionality with sample data to ensure the indexes are working as expected.
   - **Perform Manual Testing**:
     - Run searches with various filters and keywords to confirm Elasticsearch is providing fast and accurate results.

9. **Documentation**

   - **Document API Endpoints**:
     - Add endpoint descriptions, parameters, and example queries to the API documentation.

10. **Deployment and Optimization**

    - **Configure Elasticsearch for Production**:
      - Set up production configurations, including security settings, optimized indices, and backups.
    - **Monitor and Tune**:
      - Use monitoring tools to check Elasticsearch performance and make adjustments as necessary.

---

### Deliverables

- Working search functionality on the `/startups/search/` and `/projects/search/` endpoints.
- Documentation for API search endpoints.
- Tests covering indexing, querying, and filtering with Elasticsearch.
  
### Notes

https://testdriven.io/blog/django-drf-elasticsearch/ 

https://codewithmuh.medium.com/how-to-integrate-elasticsearch-with-drf-a24ede1739d2 

https://django-elasticsearch-dsl-drf.readthedocs.io/en/latest/

### PR DESCRIPTION
**Pull request by KarakaticaAsh**
Знімок екрана 2025-08-04 о 00 18 24
Знімок екрана 2025-08-05 о 00 34 12
Знімок екрана 2025-08-05 о 00 34 33
Знімок екрана 2025-08-05 о 00 37 33
Знімок екрана 2025-08-05 о 00 46 56

What I have done (with added images):

- Added Document classes for Startup and Project
- Connected post_save and post_delete signals for auto-indexing
- Implemented search views using django-elasticsearch-dsl-drf
- Added serializers for Elasticsearch documents
- Registered search endpoints in routers and core URLs
- Created unit tests for search functionality
- Updated README with basic search API usage

### CODE REVIEW

File: .env.prod [hunk 1]
```diff
@@ -0,0 +1 @@
+ELASTICSEARCH_DSL__default__hosts=https://your-prod-es-host:9200
```
Comment (mehalyna) [inline .env.prod:1]: Consider making this value configurable via deployment secrets or environment variables in your orchestration system, not hardcoded in the repo.

File: README.md [hunk 2]
```diff
@@ -169,4 +169,57 @@ Content-Type: application/json
 {
   "refresh": "<your_refresh_token>"
 }
-```
\ No newline at end of file
+```
+
+###Elasticsearch Integration
```
Comment (mehalyna) [inline README.md:174]: Add a space after ### to render the heading properly as Markdown.

File: core/settings.py [hunk 3]
```diff
@@ -69,6 +72,12 @@
     },
 }
 
+ELASTICSEARCH_DSL = {
+    'default': {
+        'hosts': 'http://localhost:9200'
+    },
+}
```
Comment (mehalyna) [inline core/settings.py:79]: Consider referencing the host from an environment variable for flexibility between dev/staging/prod. For example:
```
'hosts': os.environ.get('ELASTICSEARCH_DSL_HOST', 'http://localhost:9200')
```

File: projects/models.py [hunk 4]
```diff
@@ -50,40 +35,18 @@ class Meta:
...
         ('draft', 'Draft'),
         ('in_progress', 'In Progress'),
         ('completed', 'Completed'),
         ('cancelled', 'Cancelled')
     ]
 
+    # Якщо модель Startup в apps.startups.models
     startup = models.ForeignKey(
-        'profiles.Startup',
+        'startups.Startup',  # або 'profiles.Startup' якщо модель там
```
Comment (mehalyna) [inline projects/models.py:49]: Good update to the new app location for Startup.
Remove the comment in production code or adjust it to clarify the migration path if still in transition.

File: projects/documents.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+from django_elasticsearch_dsl import Document, Index, fields
```
Comment (mehalyna) [inline projects/documents.py:1]: You don't use `fields` here, so you can remove it from the imports.

Why use both a global `project_index` and a nested `Index` class? One is sufficient since the `Index` class inside the Document sets the index name.

File: projects/serializers.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+class ProjectDocumentSerializer(DocumentSerializer):
+    class Meta:
+        document = ProjectDocument
+        fields = (
+            "title",
+            "description",
+            "status",
+            "required_amount",
+        )
```
Comment (mehalyna) [inline projects/serializers.py:12]: Consider exposing the project id field as well for easier linking in UI.

File: projects/views.py [hunk 7]
```diff
@@ -1,12 +1,22 @@
...
+    filter_backends = [FilteringFilterBackend, CompoundSearchFilterBackend]
+    search_fields = ("title", "description")
+    filter_fields = {
+        "status": "status.raw",
+        "required_amount": "required_amount",
+    }
+    ordering_fields = {
+        "title": "title.raw",
+        "status": "status.raw",
+    }
```
Comment (mehalyna) [inline projects/views.py:22]: You may want to add pagination here if your result set can grow large.
Also, consider supporting more filter fields, e.g., by category or startup.

File: projects/urls.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 from django.urls import path, include
+from .views import ProjectSearchViewSet
 
 urlpatterns = [
   path('api/', include('users.urls')),
+  path('search/', ProjectSearchViewSet.as_view({'get': 'list'}), name='startup-search'),
```
Comment (mehalyna) [inline projects/urls.py:6]: This registers the search endpoint, but the path name suggests it's for startups rather than projects. It may be clearer as 'project-search'.

File: startups/models.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+            ('pre-seed', 'Pre-Seed'),
+            ('seed', 'Seed'),
+            ('series_a', 'Series A'),
+            ('series_b', 'Series B'),
+            ('growth', 'Growth'),
+        ],
+        default='idea'
+    )
+    location = models.CharField(max_length=255, blank=True)
+    industries = models.CharField(max_length=255, help_text="Comma-separated list of industries")
```
Comment (mehalyna) [inline startups/models.py:19]: Consider using a separate table or ManyToManyField for industries. Comma-separated fields can make querying and integrity checks harder.

File: startups/serializers.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+class StartupDocumentSerializer(DocumentSerializer):
+    class Meta:
+        document = StartupDocument
+        fields = (
+            "company_name",
+            "description",
+            "funding_stage",
+            "location",
+            "industries",
+        )
```
Comment (mehalyna) [inline startups/serializers.py:13]: Include the `id` field for reference in results.

File: startups/views.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    serializer_class = StartupDocumentSerializer
+    filter_backends = [FilteringFilterBackend, CompoundSearchFilterBackend]
+    search_fields = ("company_name", "description")
+    filter_fields = {
+        "location": "location.raw",
+        "funding_stage": "funding_stage.raw",
+    }
+    ordering_fields = {
+        "company_name": "company_name.raw",
+    }
```
Comment (mehalyna) [inline startups/views.py:21]: If you expect to filter on `industries`, consider adding it to `filter_fields`.
Also, you might want to allow ordering by `funding_stage` or location.

File: startups/urls.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+from django.urls import path
+from rest_framework.routers import DefaultRouter
+from startups.views import StartupSearchViewSet
+
+router = DefaultRouter()
+router.register(r'search', StartupSearchViewSet, basename='startup-search')
```
Comment (mehalyna) [inline startups/urls.py:6]: For consistency, consider using plural or more descriptive basenames, e.g., `startups-search`.

File: startups/tests/test_startup_elasticsearch.py
Comment (mehalyna) [inline startups/tests/test_startup_elasticsearch.py:1]: Good to see test coverage for search!

Add tests for edge cases (empty query, no results, multiple filters combined).

File: projects/tests/test_project_elasticsearch.py
Comment (mehalyna) [inline projects/tests/test_project_elasticsearch.py:1]: Good to see test coverage for search!

Add tests for edge cases (empty query, no results, multiple filters combined).

**Review by mehalyna — CHANGES_REQUESTED** [review]
Consider error handling for Elasticsearch downtime in views, with graceful fallbacks or user-friendly error messages.
Add explicit API documentation (e.g., Swagger/OpenAPI) if you want external consumers to use these endpoints.

**Review by MaksymZymyn1989 — COMMENTED** [review]
Why is StartupsConfig in projects app? Why not in startups app?

File: projects/apps.py [hunk 13]
```diff
@@ -1,6 +1,8 @@
 from django.apps import AppConfig
 
-
-class ProjectsConfig(AppConfig):
+class StartupsConfig(AppConfig):
     default_auto_field = 'django.db.models.BigAutoField'
```
[the remark this answers is not in the document]
Comment (KarakaticaAsh) [reply projects/apps.py:4]: Thank you for mention, I mismatched startups with projects apps folders.

**Comment by KarakaticaAsh** [comment]
In latest commits I considered all advises you gave me, but had an trouble with AggsProxy import, as it didn't work properly.

**Review by MaksymZymyn1989 — COMMENTED** [review]
I see now in CI/CD pipeline:
ImportError: cannot import name 'AggsProxy' from 'elasticsearch.dsl.search'

**Comment by MaksymZymyn1989** [comment]
> In latest commits I considered all advises you gave me, but had an trouble with AggsProxy import, as it didn't work properly.

Yes, I see now in CI/CD pipeline:
ImportError: cannot import name 'AggsProxy' from 'elasticsearch.dsl.search'

**Review by mehalyna — APPROVED** [review]

**Review by Raidriar28 — APPROVED** [review]
Downgrading to an earlier version seems like a reasonable solution.
